tokio/src/task/task_local.rs

```diff
@@ -299,47 +300,62 @@ impl<T: 'static> fmt::Debug for LocalKey<T> {
     }
 }
 
-/// A future that sets a value `T` of a task local for the future `F` during
-/// its execution.
-///
-/// The value of the task-local must be `'static` and will be dropped on the
-/// completion of the future.
-///
-/// Created by the function [`LocalKey::scope`](self::LocalKey::scope).
-///
-/// ### Examples
-///
-/// ```
-/// # async fn dox() {
-/// tokio::task_local! {
-///     static NUMBER: u32;
-/// }
-///
-/// NUMBER.scope(1, async move {
-///     println!("task local value: {}", NUMBER.get());
-/// }).await;
-/// # }
-/// ```
-// Doesn't use pin_project due to custom Drop.
-pub struct TaskLocalFuture<T, F>
-where
-    T: 'static,
-{
-    local: &'static LocalKey<T>,
-    slot: Option<T>,
-    future: Option<F>,
-    _pinned: PhantomPinned,
+pin_project! {
+    /// A future that sets a value `T` of a task local for the future `F` during
+    /// its execution.
+    ///
+    /// The value of the task-local must be `'static` and will be dropped on the
+    /// completion of the future.
+    ///
+    /// Created by the function [`LocalKey::scope`](self::LocalKey::scope).
+    ///
+    /// ### Examples
+    ///
+    /// ```
+    /// # async fn dox() {
+    /// tokio::task_local! {
+    ///     static NUMBER: u32;
+    /// }
+    ///
+    /// NUMBER.scope(1, async move {
+    ///     println!("task local value: {}", NUMBER.get());
+    /// }).await;
+    /// # }
+    /// ```
+    pub struct TaskLocalFuture<T, F>
+    where
+        T: 'static,
+    {
+        local: &'static LocalKey<T>,
+        slot: Option<T>,
+        #[pin]
+        future: Option<F>,
+        #[pin]
+        _pinned: PhantomPinned,
```

Comment:
I'm not really clear about whether we need this `Future` to be `!Unpin`, as making it `Unpin` seems to make life easier in some situations.

For example, if the feature in #5743 is added, it'll be common that we poll this with `&mut fut` to avoid the future being consumed, then we're able to take the value back after getting `Ready`. If the future is always `!Unpin`, we have to pin it in the heap even if the inner future is already `Unpin`. Note that pinning it on the stack doesn't work as the `Pin<&mut ..>` shadows the ownership of the future. :(

Currently I leave it untouched according to https://github.com/tokio-rs/tokio/pull/3943#discussion_r666140162.

Reply:
Removing the `PhantomPinned` field would not be breaking, but we could never undo it because _that_ would be breaking.

As for `into_value`, you make a good point. That method would not be able to take ownership of the future.